File: data-structures-and-algorithms/algorithm/leetCode/aboutArray/numberOfBoomerangs/numberOfBoomerangs.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

// Leetcode-447
// 时间复杂度：O(n^2)
// 空间复杂度：O(n)
class Solution
{
public:
          int numberOfBoomeranges(vector<pair<int, int> > &points)
          {
                    int res = 0;

                    for (int i = 0; i < points.size(); i++)
                    {
                              unordered_map<int, int> record;
                              for (int j = 0; j < points.size(); j++)
                                        if (j != i)
                                                  record[dis(points[i], points[j])]++;

                              for (unordered_map<int, int>::iterator iter = record.begin(); iter != record.end(); iter++)
                                        if (iter->second >= 2) // 可以注释该行
                                                  res += (iter->second) * (iter->second - 1);
                    }

                    return res;
          }

private:
          int dis(const pair<int, int> &pa, const pair<int, int> &pb)
          {
                    return (pa.first - pb.first) * (pa.first - pb.first) + (pa.second - pb.second) * (pa.second - pb.second);
          }
};

int main()
{

          return 0;
}

File: dataStructures&algorithms/algorithm/leetCode/aboutArray/numberOfBoomerangs/numberOfBoomerangs.php
<?php
/**
 * @date: 2022/3/14
 * @createTime: 11:24 AM
 */

// Leetcode-447
class Solution
{
    function numberOfBoomerangs($points)
    {
        $res = 0;
        $nums = count($points);
        for ($i = 0; $i < $nums; $i++) {
            $record = [];
            for ($j = 0; $j < $nums; $j++) {
                if ($j != $i) {
                    $record[$this->dis($points[$i], $points[$j])]++;
                }
            }
            foreach ($record as $value) {
                if ($value >= 2) {
                    $res += $value * ($value - 1);
                }
            }
        }
        return $res;
    }

    function dis(array $pa, array $pb)
    {
        return ($pa[0] - $pb[0]) * ($pa[0] - $pb[0]) + ($pa[1] - $pb[1]) * ($pa[1] - $pb[1]);
    }
}